### DISCUSSION TITLE
Pasting into Python REPL broken in recent brew Python builds?

### DISCUSSION THREAD

**Discussion by tzs** [Homebrew/discussions#428] [Everyday usage]
Anyone else seeing this? Thought I'd ask here before submitting an issue, in case I'm overlooking something stupid.

Sometime in the last month, pasting into the Python REPL in a terminal stopped working. At first I though it might be a Big Sur thing, but since then I've installed a fresh Mojave in a VM and installed Python3 with a freshly installed Brew, and reproduced the problem.

Here is an example, on Big Sur using Apple's Python3 to show the expected behavior, which was also the behavior of Brew Python3 previously. In the following, CTRL-V has pressed when the REPL first prompts for input, and then RETURN is pressed:

```
$ cat in
1+2
3+4
5+6
$ pbcopy < in
$ /usr/bin/python3
Python 3.8.2 (default, Nov  4 2020, 21:23:28)
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 1+2
3
>>> 3+4
7
>>> 5+6
11
>>>
>>>
```

Here is the behavior with Brew installed Python3:

```
$ type python3
python3 is /usr/local/bin/python3
$ python3
Python 3.9.1 (default, Dec 27 2020, 19:25:12)
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 1+2
3+4
5+6
  File "<stdin>", line 1
    1+2
3+4
5+6

       ^
SyntaxError: multiple statements found while compiling a single statement
>>>
```

On the fresh install of Mojave and Brew in the VM, the same problem occurs with Python 3.7 and Python 3.8, so this does not appear to be due to some upstream change in Python 3.9. (On the VM I installed and tested 3.7 first, so it is not the case that the 3.9 install messed something up that would also affect older versions).

It appears that for some reason recent Brew Python3 builds are seeing multiline pastes as a single line.

One clue to what might be going on is that if what I paste contains tabs, and I use iTerm2 as my terminal, on Apple Python3 and on older Brew installs of Python3 iTerm2 puts up a dialog asking what to do with the tabs in the paste. On recent Brew Python3 installs, it does **NOT** ask.

That suggests that the recent Brew Python builds are doing something different with the way they initialize the tty device.

**Comment by carlocab** [comment]
Strange. Are you unable to reproduce this on a non-Homebrew Python3.9?

**Reply by tzs** [reply]
I currently do not have a non-Homebrew Python 3.9. I've always used Homebrew for Python 3 so never figured out how to install it any other way. I was going to, on the assumption that the issue was something new in 3.9, but then when I tried Homebrew Python 3.7 and 3.8 on the Mojave VM and saw they had the problem too, that seemed to rule out it being anything new with 3.9.

Given that (1) Homebrew Python 3.8 was working fine a month or so ago, (2) Homebrew Python 3.7, 3.8, and 3.9 all show the problem now on a freshly installed Mojave (I also tried 3.7 on a freshly installed Big Sur with the same result), and (3) Apple Python 3.8 does not show the problem, it seems that it has to be something that changed within the last month or so in how Homebrew is building Python.

**Reply by tzs** [reply]
Update: On the Mojave VM, I downloaded the Python 3.9.1 installer from python.org and installed it.

The paste works fine there.

**Reply by carlocab** [reply]
Thanks for checking. That does suggest there may be a problem with the `python@3.9` formula. I'm also able to reproduce the behaviour described in your original post in `python@3.8` and `python@3.9`.

Can you open an issue at https://github.com/Homebrew/homebrew-core?

**Reply by tzs** [reply]
Issue opened: https://github.com/Homebrew/homebrew-core/issues/68193
